core/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2ClientAuthenticationFilter.java

```diff
@@ -16,25 +16,103 @@
 package org.springframework.security.oauth2.server.authorization.web;
 
 import org.springframework.security.authentication.AuthenticationManager;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.AuthenticationException;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.security.oauth2.server.authorization.authentication.OAuth2ClientAuthenticationToken;
+import org.springframework.util.Assert;
 import org.springframework.web.filter.OncePerRequestFilter;
-
 import javax.servlet.FilterChain;
 import javax.servlet.ServletException;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 import java.io.IOException;
 
 /**
+ * Processes a HTTP request's BASIC authorization headers, putting the result into the
+ * <code>SecurityContextHolder</code>.
+ *
+ * <p>
+ * For a detailed background on what this filter is designed to process, refer to
+ * <a href="https://tools.ietf.org/html/rfc6749>RFC 6749, Section 2.3</a>.
+ *
+ * <p>
+ * In summary, this filter is responsible for processing any request that has a HTTP
+ * request header of <code>Authorization</code> with an authentication scheme of
+ * <code>Basic</code> and a Base64-encoded <code>client_id:client_secret</code> token. For
+ * example, to authenticate client "Spring" with client secret "Secret" the following
+ * header would be presented:
+ *
+ * <pre>
+ *
+ *     Authorization: Basic U3ByaW5nOlNlY3JldA==
+ * </pre>
+ *
+ * <p>
+ * If authentication is successful, the resulting {@link Authentication} object will be
+ * placed into the <code>SecurityContextHolder</code>.
+ *
+ * <p>
+ * If authentication fails, the <code>SecurityContextHolder</code> will be cleared.
+ *
  * @author Joe Grandja
+ * @author Patryk Kostrzewa
  */
 public class OAuth2ClientAuthenticationFilter extends OncePerRequestFilter {
-	private AuthenticationManager authenticationManager;
+
+	private static final String TOKEN_ENDPOINT = "/oauth2/token";
+	private final AuthenticationManager authenticationManager;
+	private final ClientAuthenticationConverter authenticationConverter = new ClientAuthenticationConverter();
```

Comment:
Let's reference the interface `private AuthenticationConverter authenticationConverter` and provide a setter to enable customization `public final void setAuthenticationConverter(AuthenticationConverter authenticationConverter)`